import os
import pandas as pd
import glob

# set dbs and other params
GTDB = '/group/ctbrowngrp/sourmash-db/gtdb-rs226/gtdb-rs226.k21.sig.zip'
GTDB_TAX  = '/group/ctbrowngrp/sourmash-db/gtdb-rs226/gtdb-rs226.lineages.csv'
NEW_MAG_TAX = '../resources/250411_mag_taxonomy.tsv'
OWN_MAG_SIG = '/group/ctbrowngrp2/amhorst/2025-pigparadigm/results/sketches/MAGs.99.zip'
KSIZE = 21 
SCALED = 1000

OUTPUT_DIR ="/group/ctbrowngrp2/amhorst/2025-pangenome/results/test_pipeline"
MAG_LOCATION = "/group/ctbrowngrp2/amhorst/2025-pigparadigm/results/drep/drep.99/dereplicated_genomes"

# set configfile
configfile: "../config/config.yaml"
pangenome_species = config["pang_org"]
pang_name_out = config["pang_folder"]

# set list of fasta files
genomes = glob_wildcards(f"{OUTPUT_DIR}/{pang_name_out}/drep/dereplicated_genomes/{{sample}}.fasta").sample


# include snakefiles
include: "rules/collect_genomes.smk"
include: "rules/roary.smk"
include: "rules/genes_to_hash.smk"


# final output 
rule all:
    input:
        # expand(f"{OUTPUT_DIR}/{pang_name_out}/sourmash/{pang_name_out}.rankt.{KSIZE}.{SCALED}.csv"),
        expand(f"{OUTPUT_DIR}/{pang_name_out}/sourmash/{pang_name_out}.cloud.fa",),
        expand(f"{OUTPUT_DIR}/{pang_name_out}/check/drep.check"),
        expand(f"{OUTPUT_DIR}/{pang_name_out}/check/{pang_name_out}.roary.done")

